ogr/ogrsf_frmts/miramon/ogrmiramonlayer.cpp

```diff
@@ -0,0 +1,2613 @@
+
+            if (iGeom == 0)
+                bExternalRing = true;
+            else
+                bExternalRing = false;
+
+            eErr = MMDumpVertices(poNewGeometry, bExternalRing, TRUE);
+            if (eErr != OGRERR_NONE)
+                return eErr;
+        }
+    }
+    else if (eLT == wkbPoint || eLT == wkbLineString)
+    {
+        // Reads all coordinates
+        eErr = MMDumpVertices(hGeom, true, FALSE);
+
+        if (eErr != OGRERR_NONE)
+            return eErr;
+    }
+    else if (eLT == wkbGeometryCollection)
+    {
+        CPLError(
+            CE_Failure, CPLE_NotSupported,
+            "MiraMon: wkbGeometryCollection inside a wkbGeometryCollection?");
+        return OGRERR_UNSUPPORTED_GEOMETRY_TYPE;
+    }
+
+    return OGRERR_NONE;
+}
+
+/****************************************************************************/
+/*                           WriteGeometry()                                */
+/*                                                                          */
+/*      Write a geometry to the file.  If bExternalRing is true it          */
+/*      means the ring is being processed is external.                      */
+/*                                                                          */
+/****************************************************************************/
+
+OGRErr OGRMiraMonLayer::MMWriteGeometry()
+
+{
+    OGRErr eErr = MMAddFeature(phMiraMonLayer, &hMMFeature);
+
+    if (eErr == MM_FATAL_ERROR_WRITING_FEATURES)
+    {
+        CPLDebug("MiraMon", "Error in MMAddFeature() "
+                            "MM_FATAL_ERROR_WRITING_FEATURES");
+        CPLError(CE_Failure, CPLE_FileIO, "\nMiraMon write failure: %s",
+                 VSIStrerror(errno));
+        return OGRERR_FAILURE;
+    }
+    if (eErr == MM_STOP_WRITING_FEATURES)
+    {
+        CPLDebug("MiraMon", "Error in MMAddFeature() "
+                            "MM_STOP_WRITING_FEATURES");
+        CPLError(CE_Failure, CPLE_FileIO, "MiraMon format limitations.");
+        CPLError(CE_Failure, CPLE_FileIO,
+                 "Try V2.0 option (-lco Version=V2.0).");
+        return OGRERR_FAILURE;
+    }
+
+    return OGRERR_NONE;
+}
+
+/****************************************************************************/
+/*                       TranslateFieldsToMM()                              */
+/*                                                                          */
+/*      Translase ogr Fields to a structure that MiraMon can understand     */
+/****************************************************************************/
+
+OGRErr OGRMiraMonLayer::TranslateFieldsToMM()
+
+{
+    if (poFeatureDefn->GetFieldCount() == 0)
+        return OGRERR_NONE;
+
+    CPLDebug("MiraMon", "Translating fields to MiraMon...");
+    // If the structure is filled we do anything
+    if (phMiraMonLayer->pLayerDB)
+        return OGRERR_NONE;
+
+    phMiraMonLayer->pLayerDB = static_cast<struct MiraMonDataBase *>(
```

Comment:
CPLCalloc() aborts in case of memory failure, so the below test canno be true. Either remove it, or use VSICalloc() instead. Applies to the pFields allocation as well

Reply:
Done